plotly/tools.py

```diff
@@ -1529,34 +1529,46 @@ def _trisurf(x, y, z, simplices, colormap=None, color_func=None,
         import numpy as np
         from plotly.graph_objs import graph_objs
         points3D = np.vstack((x, y, z)).T
+        simplices = np.atleast_2d(simplices)
 
         # vertices of the surface triangles
         tri_vertices = points3D[simplices]
 
-        if not color_func:
+        # Define colors for the triangle faces
+        if color_func is None:
             # mean values of z-coordinates of triangle vertices
             mean_dists = tri_vertices[:, :, 2].mean(-1)
+        elif isinstance(color_func, (list, np.ndarray)):
+            # Pre-computed list / array of values to map onto color
+            if len(color_func) != len(simplices):
+                raise ValueError('If color_func is a list/array, must'
+                                 ' be the same length as simplices')
+            mean_dists = np.asarray(color_func)
         else:
             # apply user inputted function to calculate
             # custom coloring for triangle vertices
             mean_dists = []
-
             for triangle in tri_vertices:
                 dists = []
                 for vertex in triangle:
                     dist = color_func(vertex[0], vertex[1], vertex[2])
                     dists.append(dist)
-
                 mean_dists.append(np.mean(dists))
+            mean_dists = np.asarray(mean_dists)
 
-        min_mean_dists = np.min(mean_dists)
-        max_mean_dists = np.max(mean_dists)
-        facecolor = FigureFactory._map_z2color(mean_dists,
-                                               colormap,
-                                               min_mean_dists,
-                                               max_mean_dists)
-
-        ii, jj, kk = zip(*simplices)
+        # Check if facecolors are already strings and can be skipped
+        if isinstance(mean_dists[0], str):
```

Comment:
cool so then it's OK as is?

On Thu, May 26, 2016 at 6:44 PM, Adam Kulidjian <notifications@github.com>
wrote:

> In plotly/tools.py
> <https://github.com/plotly/plotly.py/pull/484#discussion_r64847443>:
>
> >
> > -        min_mean_dists = np.min(mean_dists)
> > -        max_mean_dists = np.max(mean_dists)
> > -        facecolor = FigureFactory._map_z2color(mean_dists,
> > -                                               colormap,
> > -                                               min_mean_dists,
> > -                                               max_mean_dists)
> > -
> > -        ii, jj, kk = zip(*simplices)
> > +        # Check if facecolors are already strings and can be skipped
> > +        if isinstance(mean_dists[0], str):
>
> can we check against basestring here or use six
>
> I don't know if this is relevant/has been answered, but I know for sure
> that basestring won't work between 2 and 3. str works fine though.
>
> I'll have a look tomorrow though at the rest.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/plotly/plotly.py/pull/484/files/bed542d3a268a2c469a724570ed03965903ff115#r64847443>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABwSHQyT6EFye1WR84zJeb13l3SEMAwvks5qFkyYgaJpZM4Invo8>
> .
>